### ISSUE TITLE
Token refresh does not work when USER_PASSWORD_AUTH is used with Device Tracking

### ISSUE THREAD

**Issue by pdguru81** [aws-amplify/amplify-flutter#4476]
### Description

We configured amplify flutter with the settings below. However, ID/Access Tokens are still not refreshed after they expire. It's quite strange because the docs say Amplify should do this automatically. What are we missing and how do we fix? 

### Categories

- [ ] Analytics
- [ ] API (REST)
- [ ] API (GraphQL)
- [X] Auth
- [X] Authenticator
- [ ] DataStore
- [ ] Notifications (Push)
- [ ] Storage

### Screenshots

Settings
<img width="611" alt="Screenshot 2024-02-23 at 09 38 48" src="https://github.com/aws-amplify/amplify-flutter/assets/1818047/7e863aef-4d57-46e4-9be9-8cb1c0650bf0">

Sample code we use per this [doc](https://docs.amplify.aws/flutter/start/project-setup/upgrade-guide/):

Amplify Flutter

<img width="662" alt="Screenshot 2024-02-23 at 09 39 47" src="https://github.com/aws-amplify/amplify-flutter/assets/1818047/e0478bf4-c912-4997-95df-b356b4c715b3">

### Platforms

- [X] iOS
- [X] Android
- [ ] Web
- [ ] macOS
- [ ] Windows
- [ ] Linux

### Flutter Version

3.3.10

### Amplify Flutter Version

1.6.1

### Deployment Method

Custom Pipeline

**Comment by pdguru81** [comment]
Yes, the reproduction step is pretty much what you have described. We started out with V0.X, and then migrated to V1? However, the problem persists. In our case, we end up with `SessionExpiredException` after the session expires.

**Comment by Jordan-Nelson** [comment]
> We started out with V0.X, and then migrated to V1? However, the problem persists

Okay. It sounds like you are saying that upgrading from v0 to v1 isn't required as part of the reproduction steps. I would like to confirm that. Are you able top reproduce this if you follow these steps:
1. Reset the device (on an iOS simulator you can go to Device -> Erase all Contents and Settings)
2. Install a version of the app using Amplify v1.6
3. Sign in with username and password using Amplify.signIn()
4. Call Amplify.fetchAuthSession() and observe that access/id token are valid.
5. Wait more than 5 minutes for access/id token to expire
6. Call Amplify.fetchAuthSession() and observe an exception

**Comment by pdguru81** [comment]
Yes, I am confirming that we already use V1 and don't need to upgrade. We also just tested again following step 1 to 6. After 5 minutes, what we get is  actually `SignedOutException`. This is shown in the screenshot below where we signed in using AmplifyV1.6 and then waited 5 minutes. Subsequent calls to `Amplify.fetchAuthSession() ` returned `SignedOutException`

Screenshot shows the error we log.
![Screenshot 2024-02-23 at 10 32 53 PM](https://github.com/aws-amplify/amplify-flutter/assets/1818047/5c9455c2-5968-4b65-8d70-9c9732518791)

**Comment by pdguru81** [comment]
To add, we just checked cloudtrail logs and are seeing an error about the refreshToken being invalid. It's unclear why this happens since we don't modify or manage the refresh token's directly.

```   
    "userAgent": "amplify-flutter/1.6.2 ios/17.0.1  aws-sdk-dart/0.3.1",
    "errorCode": "NotAuthorizedException",
    "errorMessage": "Invalid Refresh Token."
```

**Comment by pdguru81** [comment]
We've narrowed this issue down to the fact that device tracking prevents valid refresh tokens from being sent. Had to look at Cloudtrail logs for an idea of the problem. Why do token refreshes fail when device tracking is enabled? 🤔

**Comment by pdguru81** [comment]
I haven't heard back.  I am updating this ticket with the steps to reproduce for amplify-flutter sdk:

Steps to Reproduce
1. Create a user pool with Username and Password as the authentication method (Use email for usernames)
2. Configure a client whose ID Token and Access Token expire after 5 minutes.
3. Enable device Tracking (can set it to Opt-in)
4. Enable USER_PASSWORD_AUTH
5. Use amplify-auth. flutter sdks to connect to Cognito using the Client ID details from above
6. Sign up a new user - "TestUser". Note that the username in Cognito is "test@example.com"
7. Manually confirm the user account
8. Sign in with USER_PASSWORD_AUTH with the username as "test@example.com"
Remember the device

Login to cognito and observe that sessions don't auto-refresh after 5 minutes when device tracking is on - even in cases where refreshToken is set to expire after 30 days.

**Comment by Jordan-Nelson** [comment]
@pdguru81 Thanks for the detailed reproduction steps. I believe this is related to https://github.com/aws-amplify/amplify-flutter/issues/4501

This is a known issue with USER_PASSWORD_AUTH that we are looking into.

**Comment by Jordan-Nelson** [comment]
@pdguru81 I wanted to confirm that you are using `USER_PASSWORD_AUTH` to migrate users from a legacy user management system as described in the Cognito docs here: https://docs.aws.amazon.com/cognito/latest/developerguide/amazon-cognito-user-pools-authentication-flow.html#amazon-cognito-user-pools-user-migration-authentication-flow

**Comment by pdguru81** [comment]
Just seeing this: No, we are not migrating users from a legacy system. We have been using cognito from day 1 and started encountering this problem when we used flutter sdks to build our mobile app and manage user logins. All the users we have so far were created on cognito

**Comment by Jordan-Nelson** [comment]
@pdguru81 is there another reason that you are using the USER_PASSWORD_AUTH flow? This is typically only used when you are migrating users from another auth provider.

USER_SRP_AUTH, which is the default in Amplify, is the recommended flow by Cognito unless you have a specific use case that requires a different need. This bug seems to only be present when using USER_PASSWORD_AUTH. I believe if you could switch to USER_SRP_AUTH to resolve this.

**Comment by pdguru81** [comment]
Interesting, we will test that out and see.

**Comment by Jordan-Nelson** [comment]
@pdguru81 - Have you had a chance to test switching to the SRP flow?

**Comment by Jordan-Nelson** [comment]
@pdguru81 - Just following up on this. I wanted to see if you have had a chance to switch to the SRP flow.

**Comment by michael-joseph-payne** [comment]
I am definitely seeing similar behaviour, but have not yet confirmed that we are using USER_PASSWORD_AUTH yet.

We had been using a very old (v0) version of the flutter package, and were forced to upgrade to avoid deprecation.  Post migration, we're now receiving reports from production customers that they can not stay logged in.

We see lots of SessionExpiredException (telling us to sign in to fix) (~500 occurrences) and some NotAuthorizedExceptions (~7)

I have not found a reliable way to reproduce this, but it is definitely happening in production and is causing major headaches for our customers.  Please let me know what I can do to provide more information - this is extremely critical for us.

**Comment by Jordan-Nelson** [comment]
@michael-joseph-payne There was a known issue (see: https://github.com/aws-amplify/amplify-flutter/issues/3995) that prevented access/id tokens from being refreshed when device tracking was enabled after upgrading from v0 to v1. End users would be signed out after the access/id expired. After re-authentication tokens would be refreshed as expected. This was resolved in Amplify Flutter v1.7.2.

**Comment by michael-joseph-payne** [comment]
@Jordan-Nelson Ok, sounds good - the release in question used amplify_flutter 1.6.1 so I believe it should be addressed by upgrading to 1.7.2+

I'll check back in after we canary a bit

**Comment by michael-joseph-payne** [comment]
As a follow up question - does setting forceRefresh in FetchAuthSessionOptions work around this issue?

**Comment by Jordan-Nelson** [comment]
@michael-joseph-payne - it would not. 

The issue prior to v1.7.2 was that the device meta data was not migrated from v0 to v1. When a user signs in with device tracking enabled, tokens cannot be refreshed unless the device ID is included as part of the refresh request.

**Comment by Alvish0407** [comment]
We encounter the same issue of `SignedOutException` when invoking the `Amplify.Auth.fetchAuthSession()` method. We have configured the idToken and accessToken to expire after 1 hour, while the refreshToken is set to expire after 30 days. Interestingly, the auto-refresh mechanism works for the initial few hours, but after that, it abruptly throws the `SignedOutException`.

**Comment by ekjotmultani** [comment]
Hi @Alvish0407, thank you for adding your teams experience with this. We have recently renewed interest in this bug so we are looking into it.

**Comment by Alvish0407** [comment]
@ekjotmultani Do you believe calling the `Amplify.Auth.fetchAuthSession` multiple times within a short period could potentially cause this issue?

**Comment by ekjotmultani** [comment]
It may be, are you also enabling device tracking?

**Comment by Alvish0407** [comment]
> It may be, are you also enabling device tracking?

Nope. It's disabled.

**Comment by ekjotmultani** [comment]
@Alvish0407 Interesting, could you share how you are calling fetchAuthSession?

**Comment by Alvish0407** [comment]
We’re using WebSocket connections that require tokens to be passed on. The WebSocket automatically disconnects every two hours after the connection. We attempt to obtain the token using `Amplify.Auth.fetchAuthSession` every time before connecting to the socket, which works fine. However, the first time we connect after the two-hour interval, everything works as expected. But when we try to connect again after a total of four hours, we encounter an exception.

```shell
I/flutter (10123): 💡   "isSignedIn": false,
I/flutter (10123): 💡   "userSub": "AWSErrorResult<String, SignedOutException> {\n  \"exception\": {\n    \"message\": \"No user is currently signed in\"\n  },\n  \"stackTrace\": null\n}",
I/flutter (10123): 💡   "userPoolTokens": "AWSErrorResult<CognitoUserPoolTokens, SignedOutException> {\n  \"exception\": {\n    \"message\": \"No user is currently signed in\"\n  },\n  \"stackTrace\": null\n}",
I/flutter (10123): 💡   "credentials": "AWSSuccessResult<AWSCredentials, AuthException> {\n  \"value\": {\n    \"access_key_id\": \"*****\",\n    \"secret_access_key\": \"*****\",\n    \"token\": \"*****
I/flutter (10123): 💡   "identityId": "AWSSuccessResult<String, AuthException> {\n  \"value\": \"*****\"\n}"
```

I conducted multiple tests and consistently obtained the same outcome. I was also logging the `HubChannel.Auth` events using the `Amplify.Hub.listen` method. However, I didn’t notice any logs related to authentication events.

It’s worth noting that whenever the app experiences a cold start, we invoke `Amplify.Auth.fetchAuthSession`. Subsequently, before establishing each connection, we execute this function.

**Comment by ekjotmultani** [comment]
@Alvish0407, thank you for the great details in your response, let me take this data and investigate, I'll come back with an update when we have one (or likely more questions when I attempt to reproduce :p), I suspect this might not be related directly to the parent issue since what we uncovered before was a problem with the DeviceID not being passed properly into the InitAuth call to Cognito's API.
